pkg/utils/security_test.go

```diff
@@ -0,0 +1,162 @@
+	c.Assert(err, IsNil)
+
+	var result struct{ Path string }
+	err = tls.GetJSON("/aaa", &result)
+	c.Assert(err, IsNil)
+	c.Assert(result.Path, Equals, "/aaa")
+	err = tls.GetJSON("/bbbb", &result)
+	c.Assert(err, IsNil)
+	c.Assert(result.Path, Equals, "/bbbb")
+}
+
+func (s *securitySuite) TestGetJSONSecure(c *C) {
+	mockServer := httptest.NewTLSServer(http.HandlerFunc(respondPathHandler))
+	defer mockServer.Close()
+
+	tls := NewTLSFromMockServer(mockServer)
+
+	var result struct{ Path string }
+	err := tls.GetJSON("/ccc", &result)
+	c.Assert(err, IsNil)
+	c.Assert(result.Path, Equals, "/ccc")
+	err = tls.GetJSON("/dddd", &result)
+	c.Assert(err, IsNil)
+	c.Assert(result.Path, Equals, "/dddd")
+}
+
+func (s *securitySuite) TestInvalidTLS(c *C) {
+	tempDir := c.MkDir()
+
+	caPath := filepath.Join(tempDir, "ca.pem")
+	_, err := NewTLS(caPath, "", "", "localhost", nil)
+	c.Assert(err, ErrorMatches, "could not read ca certificate:.*")
+
+	err = ioutil.WriteFile(caPath, []byte("invalid ca content"), 0644)
+	c.Assert(err, IsNil)
+	_, err = NewTLS(caPath, "", "", "localhost", nil)
+	c.Assert(err, ErrorMatches, "failed to append ca certs")
+
+	certPath := filepath.Join(tempDir, "test.pem")
+	keyPath := filepath.Join(tempDir, "test.key")
+	_, err = NewTLS(caPath, certPath, keyPath, "localhost", nil)
+	c.Assert(err, ErrorMatches, "could not load client key pair: open.*")
+
+	err = ioutil.WriteFile(certPath, []byte("invalid cert content"), 0644)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(keyPath, []byte("invalid key content"), 0600)
+	c.Assert(err, IsNil)
+	_, err = NewTLS(caPath, certPath, keyPath, "localhost", nil)
+	c.Assert(err, ErrorMatches, "could not load client key pair: tls.*")
+}
+
+func (s *securitySuite) TestCheckCN(c *C) {
+	dir, err := os.Getwd()
+	c.Assert(err, IsNil)
+
+	dir = path.Join(dir, "tls_test")
+	caPath, certPath, keyPath := getTestCertFile(dir, "server")
+	// only allow client1 to visit
+	serverTLS, err := ToTLSConfigWithVerify(caPath, certPath, keyPath, []string{"client1"})
+	c.Assert(err, IsNil)
+
+	caPath1, certPath1, keyPath1 := getTestCertFile(dir, "client1")
+	clientTLS1, err := ToTLSConfigWithVerify(caPath1, certPath1, keyPath1, nil)
+	c.Assert(err, IsNil)
+
+	caPath2, certPath2, keyPath2 := getTestCertFile(dir, "client2")
+	clientTLS2, err := ToTLSConfigWithVerify(caPath2, certPath2, keyPath2, nil)
+	c.Assert(err, IsNil)
+
+	port := 9292
+	url := fmt.Sprintf("https://127.0.0.1:%d", port)
+	ctx, cancel := context.WithCancel(context.Background())
+	server := runServer(ctx, serverTLS, port, c)
+	defer func() {
+		cancel()
+		server.Close()
+	}()
+
+	resp, err := ClientWithTLS(clientTLS1).Get(url)
+	c.Assert(err, IsNil)
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	c.Assert(err, IsNil)
+	c.Assert(string(body), Equals, "This an example server")
+
+	// client2 can't visit server
+	_, err = ClientWithTLS(clientTLS2).Get(url)
+	c.Assert(err, ErrorMatches, ".*tls: bad certificate")
```

Comment:
will the server accept `client2` if the common name is `localhost` (the CA's common name)?

Reply:
no, you will see the log of unit test:
```
2020/03/12 12:21:44 http: TLS handshake error from 127.0.0.1:51864: client certificate authentication failed. The Common Name from the client certificate [client2 localhost] was not found in the configuration cluster-verify-cn with value: [client1]
```

the common name already include `localhost`